js/app/src/Blocks.svelte

```diff
@@ -51,6 +52,23 @@
 		create_layout
 	} = create_components();
 
+	const restore_keyed_values = (): void => {
+		let component_values_by_key: Record<string | number, ComponentMeta> = {};
+		old_components.forEach((component) => {
+			if (component.key) {
+				component_values_by_key[component.key] = component;
+			}
+		});
+		components.forEach((component) => {
+			if (component.key) {
+				const old_component = component_values_by_key[component.key];
+				if (old_component) {
+					component.props.value = old_component.props.value;
+				}
+			}
+		});
+	};
+
```

Comment:
Do you think it would be possible to do all of this in the `init` module? I think we could expose an `update_components` function where we could pass in components, and the old / new merging could be handled inside that module with all of the other component updates.

Reply:
sure, moved to init.ts

Reply:
Sorry I didn't mean just move the function but rather the old component state could be tracked in the init logic. 

We already have the previous component state. If we added an `update` method that took the new component / deps / layout, we could easily perform the logic all inside of the update method without needing to track the previous state outside of that module. 

This way we could also perhaps avoid some work on every component change and instead only do what is necessary (this may be minimal if anything).

Reply:
Gotcha, makes sense! Implemented this in https://github.com/gradio-app/gradio/pull/8110 which targets this PR.